### PR TITLE
Refactor/optimize eth_estimateGas

### PR DESCRIPTION
**Pull request by ak88**
## Changes

There is no real point in getting a perfect estimate, since the state can never be known in advance, and therefore the required gas will always vary to a certain degree.

Make our eth_estimateGas more in line with Geth by stopping, when we are within a certain margin of error. 
This margin is configurable through JsonRpcConfig. 

Before starting the binary search, also make a qualified guess on what the upper bound could be, which will dramatically reduce the amount of iterations required. 

Will also fix an issue if the lower bound is lower than intrinsic gas. 
In this case the `OutOfGasTracer` will not report it as out of gas, and the estimation will continue to set the upper bound lower despite being lower than intrinsic gas .

## Types of changes

- optimization of eth_estimateGas
- new JsonRpcConfg setting for a error margin expressed in basis points

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Requires documentation update

- [x] Yes
- [ ] No

Since this introduces a new config setting it will have to be documented. 

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

eth_estimateGas is still able to give you exact gas, if the error margin is set to 0, but the default error margin is set to 150 basispoints.

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs [hunk 1]
```diff
@@ -31,6 +36,12 @@ public class GasEstimator
 
         public long Estimate(Transaction tx, BlockHeader header, EstimateGasTracer gasTracer, CancellationToken token = new())
         {
+            return Estimate(tx, header, gasTracer, DefaultErrorMargin, token);
+        }
+        public long Estimate(Transaction tx, BlockHeader header, EstimateGasTracer gasTracer, int errorMargin, CancellationToken token = new())
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs:41]: Do we need two overloads? Just do `int errorMargin = DefaultErrorMargin`

File: src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs [hunk 2]
```diff
@@ -31,6 +36,12 @@ public class GasEstimator
...
+        {
+            if (errorMargin < 0) throw new ArgumentException("Cannot be negative.", nameof(errorMargin));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs:43]: ArgumentOutOfRangeException.ThrowIfNegative

File: src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs [hunk 3]
```diff
@@ -31,6 +36,12 @@ public class GasEstimator
...
+            if (errorMargin >= 10000) throw new ArgumentException("Expression cannot exceed 100% in basis points.", nameof(errorMargin));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs:44]: ArgumentOutOfRangeException.ThrowIfGreaterThanOrEqual

File: src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs (hunk 1 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs]: removed

File: src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs (hunk 2 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs]: done

File: src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs (hunk 3 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs]: done

**Review by LukaszRozmej — APPROVED** [review]

**Review by smartprogrammer93 — APPROVED** [review]

**Pull request merged**